feat: object store to pvc migrations (#633)
replicatedhq/kurl.sh: src/markdown-pages/install-with-kurl/removing-minio.md

---
path: "/docs/install-with-kurl/removing-object-storage"
date: "2021-12-17"
weight: 21
linktitle: "Removing Object Storage"

> title: "Removing Object Storage Dependencies"

Reviewer: @DanStough Is it okay that I changed these to the more generic "Removing Object Storage"? It felt like we were talking about removing object storage in general but then named this specifically after MinIO.

Author: Yeah, I think that's fair. I picked MinIO because I thought that would be what folks would be immediately searching for, but this suggestion is a better general description.

---
path: "/docs/install-with-kurl/removing-minio"
date: "2021-12-17"
weight: 21
linktitle: "Removing Minio"
title: "Removing Minio from your Cluster"
---

> As of [v2021.12.20-0](https://kurl.sh/release-notes/v2021.12.20-0), kURL clusters can be installed without addons having a peer dependency on object storage. 

Reviewer 2: ```suggestion
As of [v2021.12.20-0](https://kurl.sh/release-notes/v2021.12.20-0), kURL clusters can be installed without add-ons having a peer dependency on object storage. 
```

> There is also a migration path for clusters to remove object storage, such as is the case with removing Minio.

Reviewer 2: ```suggestion
There is also a migration path for clusters to remove object storage, similar to removing Minio.
```

Author: Was going for something different here. I think I wanted to say removing Minio is a specific case of removing object storage. Suggestions on better wording appreciated.

Reviewer: What about this?
```suggestion
There is also a migration path for clusters to remove object storage, whether Rook or MinIO is used for object storage.
```

Reviewer 2: I like this suggestion if that fits what you were trying to achieve Dan.

# New Installations

> The following add-ons can be installed without object storage:

Reviewer: Lots of add-ons can be installed without object storage. I like this wording since it shows that they can use object storage but now don't require it.
```suggestion
The following add-ons can now be installed with or without object storage:
```

Author: Yeah, I definitely think I missed a word here. :thumbsup:

> The following addons can be installed without object storage as noted:

Reviewer 2: ```suggestion
The following add-ons can be installed without object storage as noted:
```

> 1. **Registry**: without an object store in the installer spec, a Persistent Volume (PV) will be used for storage. Note, that if the Rook  addon is installed, the Registry will ALWAYS use this as an object store.

Reviewer 2: ```suggestion
1. **Registry**: without an object store in the installer spec, a Persistent Volume (PV) will be used for storage. 

**Note:** The Registry will always use this as an object store if the Rook  add-on is installed.
```

Author: Can we keep this on the same line? I'm pretty sure the breaks will cause re-numbering of the list elements.

Reviewer 2: ```suggestion
1. **Registry**: without an object store in the installer spec, a Persistent Volume (PV) will be used for storage. **Note:** The Registry will always use this as an object store if the Rook add-on is installed.
```

Reviewer 2: Good point! How about the above suggestion?

Reviewer: @DanStough Is this specific to Rook because people might not realize it comes with an object store?

Author: I'm going to change `Rook` to object storage, since it covers both `Rook` and `Minio`, if that works for you?

Reviewer: Yeah, that's fine.

> 1. **Velero**: without object storage in the spec, no `default` BackupStorageLocation will be created. If the `disableS3` flag is set to `true` for the KOTS add-on, a PV-backed storage location will be created as the `default` location using the [Local-Volume-Provider](https://github.com/replicatedhq/local-volume-provider) plugin.

Reviewer: Should default be in back-ticks or BackupStorageLocation? Is BackupStorageLocation a CR?

Author: `default` is the actual name of the object and BackupStorageLocation is the CR/type. @samantha-replicated what is the practice here?

Reviewer 2: General best practice is to back-tick objects and CR in-line.

> 1. **Velero**: without object storage in the spec, no `default` BackupStorageLocation will be created. If the `disableS3` flag is set to `true` for the KOTS addon, a PV-backed storage location will be created as the `default` location using the [Local-Volume-Provider](https://github.com/replicatedhq/local-volume-provider) plugin.

Reviewer 2: ```suggestion
1. **Velero**: without object storage in the spec, no `default` BackupStorageLocation will be created. If the `disableS3` flag is set to `true` for the KOTS add-on, a PV-backed storage location will be created as the `default` location using the [Local-Volume-Provider](https://github.com/replicatedhq/local-volume-provider) plugin.
```

1. **KOTS**: to allow KOTS to be deployed without an object store, the flag `kotsadm.disableS3` must be set to `true` in the installer. This will deploy KOTS as a statefulset using a Persistent Volume (PV) for storage. It will also disable the use of Minio for hostpath and NFS storage destinations.

This installer spec is an example of deploying a new cluster without using any object storage.
```yaml
apiVersion: cluster.kurl.sh/v1beta1
kind: Installer
metadata:
  name: no-object-storage
spec:
  kubernetes:
    version: 1.21.x
  containerd:
    version: 1.4.x
  weave:
    version: 2.6.5
  longhorn:
    version: 1.2.x
  registry:
    version: 2.5.7
  velero:
    version: 1.7.x
  kotsadm:
    version: 1.58.x
    disableS3: true
```

> # Existing Cluster Migrations

Reviewer: Could this confuse someone since "existing cluster" means something else in Replicated? Should we say pre-existing or something like that to disambiguate?

Author: I like to defer to the simpler choice of `cluster migrations` if that works for you.

Reviewer: Works for me!

> There are two migration paths for removing object storage components.

Reviewer: Splitting this into two sections doesn't feel right to me. If you look at the three add-ons in each of the next two sections, we repeat a lot of info. For example, under removing the provider we talk about how KOTS will use a StatefulSet and migrate, but that isn't really because of the removal of the provider, that's because of the `disableS3` flag. I wonder if this would be clearer if we just had one section that talks about both actions (`disableS3: true` and removing the provider), and then for each action taken, we say which of those two is responsible for it. I can write it up, I just wanted to suggest it first to see if it makes more sense that way.

Author: I understand your point. I think my intent on making it into two sections was that there are two buckets of vendors in my mind:
1. Those that will enable this flag, but for some reason or another NOT remove Rook. Why? I don't know, but IICR some apps depend on the Rook RGW. Maybe this is also some migration path for them to mitigate a bunch of changes to their installer.
1. Those that will do everything in their power to remove Rook/Minio/Object Storage. 

If you would still like to change, I'm onboard.

## Enabling `disableS3` in the KOTS Addon

> Object storage can be removed from some of the addons in the cluster by simply setting the `kotsadm.disableS3` flag in the KOTS addon to `true`.

Reviewer 2: ```suggestion
Object storage can be removed from some of the add-ons in the cluster by simply setting the `kotsadm.disableS3` flag in the KOTS addon to `true`.
```

> See documentation in the [KOTS addon](/docs/add-ons/kotsadm) for more information on the `disableS3` flag.

Reviewer 2: ```suggestion
See the documentation in the [KOTS add-on](/docs/add-ons/kotsadm) for more information on the `disableS3` flag.
```

When you re-install or upgrade using the updated installer spec (see the previous section for a sample), you should expect:
1. **Registry**: Nothing will change about the deployment. The registry will continue to use the same object storage.

> 1. **Velero**: the `default` BackupStorageLocation will be updated to point at an attached Persistent Volume using the [Local-Volume-Provider](https://github.com/replicatedhq/local-volume-provider) plugin. A migration will be triggered to copy from the object store into the attached PV.

Reviewer 2: ```suggestion
1. **Velero**: the default `BackupStorageLocation` will be updated to point at an attached Persistent Volume using the [Local-Volume-Provider](https://github.com/replicatedhq/local-volume-provider) plugin. A migration will be triggered to copy from the object store into the attached PV.
```

Author: `default` is the name name of the item and `BackupStorageLocation` is the type. I'll change them both to monospace.

1. **KOTS**: This will scale down the KOTS deployment, delete it, and re-deploy KOTS as a statefulset using a Persistent Volume (PV) for storage. It will also disable the use of Minio for hostpath and NFS storage destinations. A migration will be triggered to copy from the object store into the attached PV.

## Removing the Existing Provider

To fully remove object storage from the cluster, the current provider must be removed from your installer spec.

> In the case of Minio, it is a straightforward removal of the addon.

Reviewer 2: ```suggestion
In the case of Minio, it is a straightforward removal of the add-on.
```

> For clusters using the Rook adddon, another CSI such as Longhorn or OpenEBS is required for storage. 

Reviewer 2: ```suggestion
For clusters using the Rook addd-on, another CSI such as Longhorn or OpenEBS is required for storage. 
```

Data will be migrated between the storage providers automatically per existing [CSI Migrations](/docs/install-with-kurl/migrating-csi).

> Any object store data used by addons will also be migrated to PVs.

Reviewer 2: ```suggestion
Any object store data used by add-ons will also be migrated to PVs.
```

You are also expected to set the `kotsadm.disableS3` flag to `true` in the installer to explicitly enable migrations for KOTS.

> See documentation in the [KOTS addon](/docs/add-ons/kotsadm) for more information on the `disableS3` flag.

Reviewer 2: ```suggestion
See the documentation in the [KOTS add-on](/docs/add-ons/kotsadm) for more information on the `disableS3` flag.
```
